File: apps/frontend/src/services/trends.ts
import api, { ApiResponse, unwrap } from "./api";

export interface TrendFilters {
  state?: string;
  district?: string;
  mandi?: string;
  commodity?: string;
  variety?: string;
  grade?: string;
  date_from?: string;
  date_to?: string;
}

export interface TrendPoint {
  date: string;
  avg_price: number;
  arrivals_qtl: number;
}

export interface TrendWindow {
  window_days: number;
  points: TrendPoint[];
  change_pct: number;
  average_price: number;
  volatility: number;
}

export interface SeasonalComparisonItem {
  season: string;
  average_price: number;
  average_arrivals_qtl: number;
  count: number;
}

export interface PriceSpikeAlert {
  date: string;
  change_pct: number;
  change_abs: number;
  note: string;
}

export interface TrendAnalyticsResponse {
  filters: TrendFilters;
  windows: TrendWindow[];
  seasonal: SeasonalComparisonItem[];
  volatility: number;
  alerts: PriceSpikeAlert[];
  generated_at: string;
  cached?: boolean;
}

export const fetchTrendAnalytics = async (filters: TrendFilters): Promise<TrendAnalyticsResponse> => {
  const response = await api.get<ApiResponse<TrendAnalyticsResponse>>("/analytics/trends", { params: filters });
  return unwrap(response.data);
};

File: apps/backend/app/schemas/trends.py
from __future__ import annotations

from datetime import date, datetime
from typing import List, Optional

from pydantic import BaseModel


class TrendFilters(BaseModel):
    state: Optional[str] = None
    district: Optional[str] = None
    mandi: Optional[str] = None
    commodity: Optional[str] = None
    variety: Optional[str] = None
    grade: Optional[str] = None
    date_from: Optional[date] = None
    date_to: Optional[date] = None


class TrendPoint(BaseModel):
    date: date
    avg_price: float
    arrivals_qtl: float


class TrendWindow(BaseModel):
    window_days: int
    points: List[TrendPoint]
    change_pct: float
    average_price: float
    volatility: float


class SeasonalComparisonItem(BaseModel):
    season: str
    average_price: float
    average_arrivals_qtl: float
    count: int


class PriceSpikeAlert(BaseModel):
    date: date
    change_pct: float
    change_abs: float
    note: str


class TrendAnalyticsResponse(BaseModel):
    filters: TrendFilters
    windows: List[TrendWindow]
    seasonal: List[SeasonalComparisonItem]
    volatility: float
    alerts: List[PriceSpikeAlert]
    generated_at: datetime
    cached: bool = False
